### ISSUE TITLE
`vim.lsp.enable()`: do not launch LSP server if root dir not found (like `single_file_support=false` from nvim-lspconfig)

### ISSUE THREAD

**Issue by andrewbraxton** [neovim/neovim#31762]
### Problem

Many LSPs can initialize without a root dir, but their usefulness may be heavily limited. For instance, without a compile_commands.json, clangd is quite helpless. In such a case it may be preferred to not have the language server active at all, as their presence is more distracting than helpful.

### Expected behavior

`vim.lsp.enable()` could be made to respect a new flag in `vim.lsp.Config` that works similarly to `single_file_support` from `nvim-lspconfig`. This flag would tell `vim.lsp.enable()` to only launch the language server when one of the `root_markers` specified in `vim.lsp.Config` is found.

**Comment by justinmk** [comment]
> a new flag in `vim.lsp.Config` that works similarly to `single_file_support` from `nvim-lspconfig`. This flag would tell `vim.lsp.enable()` to only launch the language server when one of the `root_markers` specified in `vim.lsp.Config` is found.

Er... I thought the root dir is found by root markers. Yet this issue title says "without a root dir"?

Mentioning `single_file_support` seems confusing since it sounds like you are requesting a way to NOT activate the LSP server.

**Comment by andrewbraxton** [comment]
> you are requesting a way to NOT activate the LSP server.

Correct, that's what `single_file_support = false` does. It says "if none of the root markers are found (i.e. there is no root dir), do not launch the server."

The current behavior of `vim.lsp.enable()` is to launch the server with a nil root dir if no root marker is matched, similar to `single_file_support = true`.

Perhaps the new title is more clear?

**Comment by mfussenegger** [comment]
Workaround: Don't use `vim.lsp.enable` but `vim.lsp.start` manually.

**Comment by andrewbraxton** [comment]
Yes, that's what I'm doing. I'd like to move to vim.lsp.enable().

**Comment by justinmk** [comment]
> I'd like to move to vim.lsp.enable(), since my understanding is that its purpose is to replace the autocommand+vim.lsp.start() boilerplate.

 https://github.com/neovim/neovim/pull/31771  should allow avoiding boilerplate (except the "autocommand" part), by passing the result of `vim.lsp.config[<name>]` to start().

**Comment by lewis6991** [comment]
I think this is reasonable enough to support. Lspconfig already had first class support for it, and it seems all we need to add is a single boolean field to `vim.lsp.Config` to indicate whether we should attach if no root dirs can be found. This Boolean can then be specified by the user in the `*` config to apply globally, and then also set in specific server configs for servers that don't support it.

I'm assuming the current default (effectively `single_file_support=true`) should remain?

**Comment by justinmk** [comment]
> it seems all we need to add is a single boolean field to `vim.lsp.Config` to indicate whether we should attach if no root dirs can be found.

Can we just document a trivial function that the user can set on `root_dir` ?

**Comment by lewis6991** [comment]
Don't think so. I don't think it's currently possible to use `vim.lsp.enable` with `single_file_support=false`. We need to add code to avoid calling `vim.lsp.start` if there is no resolved root dir.

**Comment by andrewbraxton** [comment]
As far as defaults, lspconfig has 183 LSPs with `single_file_support = true` as their default and 5 with `false`. So I think keeping the current default makes sense.

I'm guessing lspconfig chose the default based on whether the LSP works at all; I assume those 5 LSPs completely break without a rootdir. This new flag would allow the user to handle both this case (LSP doesn't work at all without a rootdir) and the case I described originally (LSP can successfully launch but would not be useful, like clangd).

**Comment by pawelgrzybek** [comment]
I am looking forward to this feature.

Another use case of this feature, is when there two LSP servers targeting the same filetype, but they collide with each other. A good example is a `.ts` (typescript file) and TypeScript and Deno LSPs both enabled at the same time.

I am glad to see this one coming soon 🙏

**Comment by justinmk** [comment]
`workspace_required` will be included in Nvim 0.11.1
